plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java

```diff
@@ -272,8 +277,23 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Versio
             public void onResponse(T response) {
                 logCooldownInfo();
                 final Scheduler.Cancellable existing = finalizationFuture.getAndSet(
-                    threadPool.schedule(ActionRunnable.wrap(wrappedListener, l -> l.onResponse(response)),
-                        coolDown, ThreadPool.Names.SNAPSHOT));
+                        threadPool.schedule(ActionRunnable.wrap(wrappedListener, l -> clusterService.submitStateUpdateTask(
```

Comment:
what happens on master failover here?

Is there still clean-up happening? Are we still finalizing, but without considering the delay?

Reply:
> what happens on master failover here?

The delete/snapshot entry is still in the CS.
In the case of a delete: Things will be retried on the next master, depending on whether or not the index-N blob was already updated the delete will either go through again cleanly (if it wasn't yet updated) or fail (but will still wait). In this case the removal of the delete from the CS happens in the error handler in `SnapshotsService`.
In case of the snapshot finalization: Finalization will happen again and cleanly (because the mechanics of the pending generation make sure we don't run into any generation conflicts) on the new master. It will happen without delay on the new master (not a problem, we only need the delay for the shard level index-N blobs, at the root we're tracking N via the CS) but the new master will then add the delay after the snapshot.

=> I think we're once against safe in all cases here.